lib/fluent/plugin/compressable.rb

```diff
@@ -0,0 +1,89 @@
+#
+# Fluentd
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'zlib'
+
+module Fluent
+  module Plugin
+    module Compressable
+      def compress(data, **kwargs)
+        output_io = kwargs[:output_io]
+        io = output_io || StringIO.new
+        Zlib::GzipWriter.wrap(io) do |gz|
+          gz.write data
+        end
+
+        output_io || io.string
+      end
+
+      # compressed_data is String like `compress(data1) + compress(data2) + ... + compress(dataN)`
+      # https://www.ruby-forum.com/topic/971591#979503
+      def decompress(compressed_data = '', **kwargs)
+        output_io = kwargs[:output_io]
+        input_io = kwargs[:input_io]
+
+        case
+        when input_io && output_io
+          io_decompress(input_io, output_io)
+        when compressed_data.empty?
```

Comment:
This code checks the size of compressed_data(string) is 0 or not.

becuase `Zlib::GzipReader.new` raises Error, if `Zlib::GzipReader.new` recieved  `io` like `StringIO.new('')` , 
https://github.com/fluent/fluentd/pull/1172/files#diff-a36e94f98cbad7c4380816a06392d343R57